asm/instruction.go

```diff
@@ -299,16 +339,111 @@ func (ins Instruction) Format(f fmt.State, c rune) {
 	}
 
 ref:
-	if ins.Reference != "" {
-		fmt.Fprintf(f, " <%s>", ins.Reference)
+	if ins.Reference() != "" {
+		fmt.Fprintf(f, " <%s>", ins.Reference())
 	}
 }
 
+func (ins Instruction) equal(other Instruction) bool {
+	return ins.OpCode == other.OpCode &&
+		ins.Dst == other.Dst &&
+		ins.Src == other.Src &&
+		ins.Offset == other.Offset &&
+		ins.Constant == other.Constant
+}
+
 // Size returns the amount of bytes ins would occupy in binary form.
 func (ins Instruction) Size() uint64 {
 	return uint64(InstructionSize * ins.OpCode.rawInstructions())
 }
 
+// SetReference set a reference(e.g. a jump) to another symbol
+func (ins *Instruction) SetReference(ref string) {
+	if (ins.metadata != nil && ins.metadata.reference == ref) ||
+		(ins.metadata == nil && ref == "") {
+		return
+	}
+
+	ins.metadata = ins.metadata.copy()
+	ins.metadata.reference = ref
+}
+
+// Reference denotes a reference (e.g. a jump) to another symbol.
+func (ins Instruction) Reference() string {
+	if ins.metadata == nil {
+		return ""
+	}
+
+	return ins.metadata.reference
+}
+
+// WithContext adds information about the origin/source of the instruction.
+func (ins Instruction) WithContext(ctx fmt.Stringer) Instruction {
```

Comment:
> What about `WithInfo`, from extInfos? This is still reminiscent of the original term.

I don't mind that, this came about because we would like to define a comment for example which is not related to extInfos as such.

> Also, I think this could just be a string; a `SimpleContext` feels a bit awkward.

We discussed having just a string with a custom format e.g. `foo.c:12:34: {line or comment}` with optional parts so a plain string `These instructions perform ### function.` or `firewall.go:123: asm.Ja.Label("some-label")`. Downside is that this is somewhat less flexible than defining a interface, and @lmb was concerned about the the code size/performance implication of a string parser.

> Thinking about marshaling this later, if an instruction has an Info and is a Symbol, we emit a funcInfo

We also need a BTF type ID to creating funcInfo, which we currently don't have metadata for. I wanted to implement that when we start work on marshalling.

> The kernel requires there be exactly one funcInfo pointing at the start of each subprog

I don't think that the kernel requires this. Perhaps in some cases, but I have loaded programs without BTF info before without issue.

> iirc an instruction cannot have both a lineInfo and a funcInfo

Not the case, I have seen a few programs which have both at the first instruction.

Reply:
> I don't think that the kernel requires this. Perhaps in some cases, but I have loaded programs without BTF info before without issue.

Loading programs without BTF has always been possible (BTF is relatively recent), but if a BTF blob is present, each subprog must be accompanied by a valid funcInfo with the right offset.

> Not the case, I have seen a few programs which have both at the first instruction.

Yep, then we need an extra field. I forgot that funcInfo is actual type info of `func -> func_proto` that also contains the args and their types, so we can't just synthesize it from the Symbol() value. :/ Let's take this into account before settling on the `WithContext` name, let me mull it over for a bit. LineInfo and FuncInfo are both 'contextual' information, I still find the term 'context' too generic for what it actually describes.

Maybe something like `WithLineInfo()` and `WithFuncInfo()` would be more clear? An important use case is checking whether the instruction contains a FuncInfo, so a single func/line Stringer field makes this rather difficult from the caller pov.

Any ideas?

Reply:
> subprog must be accompanied by a valid funcInfo with the right offset.

Yea, discovered that the hard way recently, a good reason to add funcInfo before moving on the generating BTF. I already have something brewing for that. But adding lineInfo without the func info doesn't hurt until that point.

>  LineInfo and FuncInfo are both 'contextual' information, I still find the term 'context' too generic for what it actually describes.

Agreed, I think `WithLineInfo` is good, that is after all exactly what this is, information about a line of source code. And while we allow any stringer to serve as the actual `Line` part of it for the sake of comments, the name still fits, and doesn't clash with funcInfo.